File: Packets/Packet.vb
Partial Public MustInherit Class Packet
    Sub New(Optional ByVal Bytes() As Byte = Nothing, Optional ByVal ForcesVLQRecognition As Boolean = False, Optional ByVal DoubleLength As Boolean = False)
        If Bytes Is Nothing Then Exit Sub
        If Bytes.Length = 0 Then Exit Sub
        RawBytes = Bytes
        OPCode = ReadByte(True)
        Dim length As Integer = ReadByte(True)
        If DoubleLength Then length /= 2
        index = 0
        If length <> Bytes.Length - 2 Or ForcesVLQRecognition Then
            'sVLQ packet
            index += 1
            length = ReadsVLQ(True)
            If length < 0 Then
                'zlib compressed
                Payload = Misc.Deflate(BasicUFL.ToList(Bytes).GetRange(index, Bytes.Count - index).ToArray)
            Else
                'uncompressed
                Payload = BasicUFL.ToList(Bytes).GetRange(index, Bytes.Count - index).ToArray
            End If
        Else
            'Non sVLQ packet
            Payload = BasicUFL.ToList(Bytes).GetRange(2, Bytes.Count - 2).ToArray
        End If
        index = 0
    End Sub
    Public Overridable Property OPCode As Byte
    Public Property RawBytes As Byte()
    Public Property Payload As Byte()
    Public MustOverride Function GetByteArray() As Byte()
    'Public MustOverride Sub LoadByteArray(ByVal ByteArray As Byte())
    Private index As UInteger = 0
    Public Function Package(Optional ByVal UsesVLQLengthing As Boolean = False, Optional ByVal DoubleLength As Boolean = False) As Byte()
        Dim size
        Dim mult
        If DoubleLength Then
            mult = 2
        Else
            mult = 1
        End If
        Dim f As Boolean
        If (Payload.Count * mult) > 256 Or UsesVLQLengthing = True Then
            f = True
            size = TosVLQ(Payload.Count * mult)
        Else
            f = False
            size = CByte(Payload.Count * mult)
        End If
        If Not f Then
            Dim ar As Byte() = {OPCode, size}
            Return Combine({ar, Payload})
        Else
            Dim ar As Byte() = {OPCode}
            Return Combine({ar, size, Payload})
        End If
    End Function
    Public Function ReadByte(Optional ByVal Raw As Boolean = False) As Byte
        Dim r
        If Raw Then
            r = RawBytes(index)
        Else
            r = Payload(index)
        End If
        index += 1
        Return r
    End Function
    Public Function ReadByteArray() As Byte()
        Return ReadBytes(ReadByte())
    End Function
    Public Function ReadBytes(ByVal Amount As UInteger) As Byte()
        Dim r As New List(Of Byte)
        For i As UInteger = 1 To Amount
            r.Add(Payload(index))
            index += 1
        Next
        Return r.ToArray
    End Function
    Public Function ReadUInt32() As UInt32
        Return BitConverter.ToUInt16(ReadBytes(4), 0)
    End Function
    Public Function ReadString(Optional ByVal Factor As Single = 1) As String
        Dim len = ReadByte() * Factor
        If len < 1 Then Return ""
        Dim l As New List(Of Byte)
        For i = 1 To len
            l.Add(ReadByte)
        Next
        Return System.Text.Encoding.UTF8.GetChars(l.ToArray)
    End Function
    Public Function ReadVLQ(Optional ByVal Raw As Boolean = False) As Integer
        Dim l As New List(Of Byte)
        Dim b As Boolean = False
        Dim i = index
        Do Until b Or i = RawBytes.Count
            l.Add(ReadByte(Raw))
            If (New Binary(l.Last)).IsLastVLQByte Then b = True
        Loop
        If b = False Then Throw New IndexOutOfRangeException("No VLQ (end) found")
        Return VLQ.FromVLQ(l.ToArray)
    End Function
    Public Function ReadsVLQ(Optional ByVal Raw As Boolean = False) As Integer
        Dim l As New List(Of Byte)
        Dim b As Boolean = False
        Dim i = index
        Do Until b Or i = RawBytes.Count
            l.Add(ReadByte(Raw))
            If (New Binary(l.Last)).IsLastVLQByte Then b = True
        Loop
        If b = False Then Throw New IndexOutOfRangeException("No VLQ (end) found")
        Return VLQ.FromsVLQ(l.ToArray)
    End Function
End Class

File: Packets/Packet.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
public abstract partial class Packet
{
	public Packet(byte[] Bytes = null, bool ForcesVLQRecognition = false)
	{
		if (Bytes == null)
			return;
		if (Bytes.Length == 0)
			return;
		RawBytes = Bytes;
		OPCode = ReadByte(true);
		int length = ReadByte(true);
		index = 1;
		if (length != Bytes.Length - 2 | ForcesVLQRecognition) {
			//sVLQ packet
			length = (int)ReadsVLQ(true);
			if (length < 0) {
				//zlib compressed
                List<byte> metapayload = Misc.ToList(Bytes);
                Payload = Misc.Deflate(metapayload.GetRange(index, Bytes.Length - index).ToArray());
                isCompressed = true;
			} else {
				//uncompressed
                Payload = Misc.ToList(Bytes).GetRange(index, Bytes.Length - index).ToArray();
			}
		} else {
			//Non sVLQ packet
            Payload = Misc.ToList(Bytes).GetRange(2, Bytes.Length - 2).ToArray();
		}
		index = 0;
	}
	public virtual byte OPCode { get; set; }
    public bool isCompressed = false;
	public byte[] RawBytes { get; set; }
	public byte[] Payload { get; set; }
	public abstract byte[] GetByteArray();
	//Public MustOverride Sub LoadByteArray(ByVal ByteArray As Byte())
	private int index = 0;
	public byte[] Package(bool UsesVLQLengthing = false, bool DoubleLength = false)
	{
		dynamic size = null;
		dynamic mult = null;
		if (DoubleLength) {
			mult = 2;
		} else {
			mult = 1;
		}
		bool f = false;
        if ((Payload.Length * mult) > 256 | UsesVLQLengthing == true)
        {
			f = true;
            size = VLQ.TosVLQ(Payload.Length * mult);
		} else {
			f = false;
            size = Convert.ToByte(Payload.Length * mult);
		}
		if (!f) {
			byte[] ar = {
				OPCode,
				size
			};
            byte[][] lel = {ar,	Payload	};
            return Misc.Combine(lel);
		} else {
			byte[] ar = { OPCode };
            byte[][] lel = {ar,	size,Payload};
			return Misc.Combine(lel);
		}
	}
	public byte ReadByte(bool Raw = false)
	{
		dynamic r = null;
		if (Raw) {
			r = RawBytes[index];
		} else {
			r = Payload[index];
		}
		index += 1;
		return r;
	}
	public byte[] ReadByteArray(bool signed = false)
	{
        if (signed)
        {
            return ReadBytes((ulong)ReadsVLQ());
        }
        else
        {
            return ReadBytes(ReadVLQ());
        }
	}
	public byte[] ReadBytes(ulong Amount)
	{
		List<byte> r = new List<byte>();
		for (uint i = 1; i <= Amount; i++) {
			r.Add(Payload[index]);
			index += 1;
		}
		return r.ToArray();
	}
	public UInt32 ReadUInt32()
	{
		return BitConverter.ToUInt16(ReadBytes(4), 0);
	}
	public string ReadString(double Factor = 1)
	{
		dynamic len = ReadByte() * Factor;
		if (len < 1)
			return "";
		List<byte> l = new List<byte>();
		for (int i = 1; i <= len; i++) {
			l.Add(ReadByte());
		}
		return new string(System.Text.Encoding.UTF8.GetChars(l.ToArray()));
	}
	public ulong ReadVLQ(bool Raw = false)
	{
		List<byte> l = new List<byte>();
		bool b = false;
		dynamic i = index;
		while (!(b | i == RawBytes.Length))
        {
			l.Add(ReadByte(Raw));
			if ((new Binary().NewByte(l[l.Count - 1])).IsLastVLQByte())
				b = true;
		}
		if (b == false)
			throw new IndexOutOfRangeException("No VLQ (end) found");
		return VLQ.FromVLQ(l.ToArray());
	}
	public long ReadsVLQ(bool Raw = false)
	{
		List<byte> l = new List<byte>();
		bool b = false;
		dynamic i = index;
        while (!(b | i == RawBytes.Length))
        {
			l.Add(ReadByte(Raw));
            if ((new Binary().NewByte(l[l.Count - 1])).IsLastVLQByte())
				b = true;
		}
		if (b == false)
			throw new IndexOutOfRangeException("No VLQ (end) found");
		return VLQ.FromsVLQ(l.ToArray());
	}
    public dynamic ReadVariant()
	{
	  int varType = ReadByte();
	  
	  switch(varType)
	  {
        case 0x01:
          return null;
	    case 0x02:
          //double
          throw new NotImplementedException("Doubles aren't implemented as variants yet");
	    case 0x03:
	      //bool
          return ReadBoolean();
	    case 0x04:
	      //signed vlq
	      return ReadsVLQ();
	    case 0x05:
	      //string
	      return ReadString();
        case 0x06:
	      //variant
	      //return ReadVariant
	      uint varSize = ReadUInt32();
	      List<dynamic> varList = new List<dynamic>(); //[varSize];
	      for(int i = 0; i < varList.Count; i++)
	      {
	      	varList.Add(ReadVariant());
	      }
          return varList;
	    case 0x07:
	      //dict variant (map)
	      return ReadVariantDict();
        default:
          throw new Exception("ReadVariant failed!");
	  }
	}

    public Dictionary<string, int> ReadVariantDict(bool Raw = false)
	{
	  Dictionary<string, int> varDict = new Dictionary<string, int>();
	  UInt32 varSize = ReadUInt32();
	  try
	  {
          for (int i = 0; i < varSize; i++)
	    {
	      varDict.Add(ReadString(i), ReadVariant());
	    }
	    return varDict;
	  }
	  catch
	  {
	    throw new Exception("ReadVariant failed!");
	  }
	}

    public bool ReadBoolean()
    {
        return Convert.ToBoolean(ReadByte());
    }
}
